scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -160,41 +188,112 @@
+    ;       kubernetes states that are not expected to occur, while pod-has-just-completed is invoked in all of the
+    ;       normal exit states (including exit-with-failure). I.e., these MUST only be invoked if expected state is not
+    ;       in a terminal state. These functions return new expected-state-dict's so should be invoked last in almost
+    ;       all cases. (However, in a few cases, (delete-task ? ?) can be invoked afterwards. It deletes the key from
+    ;       the expected state completely. Should only be invoked if we're nil.)
+    ;
+    ;   weird can be called at any time to indicate we're in a wierd state and extra logging is warranted.
+    ;
+    ;   We always end with one of  pod-has-just-completed/pod-was-killed/pod-has-started or delete-task.
+    ;
+    ;   We always update datomic first (with pod-has-* pod-was-*), etc, then we update kubernetes so we can handle restarts.
+    ;
+    ;   (delete-task ? ?) is only valid when kubernetes is in a terminal-state.
+    ;
+    ;   Note that we have release semantics. We could fail to process any operation. This means that, where relevant, we
+    ;     first write to datomic, then change into the next step. We thus show some intermediate states in the machine
+    ;     that seem like they could be skipped.
+    ;
+    ;   In the (:completed,*) states, we delete from datomic, transitioning into (:completed,:missing) and thence to (:missing, :missing)
+    ;     When we're in a terminal pod state.
+    ;
+    ;   We only delete a pod if and only if the pod is in a terminal (or unknown) state.
+    ; In some cases, we may mark mea culpa retries (if e.g., the node got reclaimed, so the pod got killed). In weird cases, we always need to mark mea culpa retries.
     (let
       [new-expected-state-dict (case (vector (or expected-state :missing) (or (:state synthesized-state) :missing))
                                  [:expected/starting :missing] (launch-task api-client expected-state-dict)
                                  [:expected/starting :pod/running] (pod-has-started kcc existing-state-dict)
-                                 ; TODO We need to add cases for the other [:expected/starting *] states in.
-                                 [:expected/starting :pod/waiting] expected-state-dict
-                                 [:expected/starting :pod/succeeded] (pod-has-just-completed kcc existing-state-dict)
-                                 [:expected/starting :pod/failed] (pod-has-just-completed kcc existing-state-dict)
+                                 [:expected/starting :pod/waiting] expected-state-dict ; Its starting. Can be stuck here.
+                                 [:expected/starting :pod/succeeded] (pod-has-just-completed kcc existing-state-dict) ; Finished fast.
+                                 [:expected/starting :pod/failed] (pod-has-just-completed kcc existing-state-dict) ; TODO: May need to mark mea culpa retry
```

Comment:
In the current state of the code how do we know that it was a mea-culpa failure? Is that going to be addressed later? If so, should the command drop the `May`?

Reply:
How do we distinguish between these various failures in the current state diagram? We would need that to support mea-culpa failures.

Reply:
I'm thinking they'd be tracked as an additional argument to pod-has-just-completed (in some cases) and for explicitly killed tasks, (if this feature is needed), they would be done via an additional key in the expected state.